unblob/handlers/compression/lzh.py

```diff
@@ -0,0 +1,102 @@
+import io
+from typing import List, Optional
+
+from structlog import get_logger
+
+from ...file_utils import Endian
+from ...models import StructHandler, ValidChunk
+
+logger = get_logger()
+
+PADDING_LEN = 2
+LEVEL_IDENTIFIERS = [0, 1, 2]
+METHOD_IDS = [
+    b"-lh0-",
+    b"-lzs-",
+    b"-lz4-",
+    b"-lh1-",
+    b"-lh2-",
+    b"-lh3-",
+    b"-lh4-",
+    b"-lh5-",
+    b"-lh6-",
+    b"-lh7-",
+    b"-lh8-",
+    b"-lhd-",
+]
+
+
+class LZHHandler(StructHandler):
+
+    NAME = "lzh"
+
+    YARA_RULE = r"""
+        strings:
+            // 2 header bytes followed by method ID
+            $magic = /[\x00-\xff][\x00-\xff](-lh0-|-lzs-|-lz4-|-lh1-|-lh2-|-lh3-|-lh4-|-lh5-|-lh6-|-lh7-|-lh8-|-lhd-)/
+        condition:
+            $magic
+    """
+
+    C_DEFINITIONS = r"""
+        struct lzh_default_header {
+            uint8 header_size;          // excludes extended headers size
+            uint8 header_checksum;
+            char method_id[5];
+            uint32 compressed_size;     // includes all extended headers size (if level 1)
+            uint32 uncompressed_size;
+            uint32 timestamp;
+            uint8 fd_attribute;
+            uint8 level_identifier;
+        };
+
+        struct level_2_header {
+            uint16 header_size; // includes all extended headers
+            char method_id[5];
+            uint32 compressed_size;     // excludes all extended headers
+            uint32 uncompressed_size;
+            uint32 timestamp;
+            uint8 fd_attribute;
+            uint8 level_identifier;
+        };
+    """
+    HEADER_STRUCT = "lzh_default_header"
+
+    def calculate_chunk(
+        self, file: io.BufferedIOBase, start_offset: int
+    ) -> Optional[ValidChunk]:
+
+        header = self.parse_header(file, Endian.LITTLE)
+        file.seek(-len(header), io.SEEK_CUR)
+
+        if (
+            header.method_id not in METHOD_IDS
+            or header.level_identifier not in LEVEL_IDENTIFIERS
+        ):
+            return
+
+        if header.level_identifier == 0x2:
+            # magic trick, just lookup the header definition of level 2
+            header_size = header.header_size + header.header_checksum << 8
```

Comment:
I am sure this is only issue for me, because I am soooo not used to do bitwise operations .... but I wanted to make this whole `if` block simpler, and could not ....  (kudos for you for creating cool integration tests!!!)

.... I just found out that the `+` operator has higher dependency than the `<<` :sweat_smile: 

So just for the sake of dumb people like me, could you use parenthesis like this?
```suggestion
            header_size = (header.header_size + header.header_checksum) << 8

```

Reply:
Ahah, nice catch ! It worked because the size was lower than `0xff` for all my test files, what I really want to do is this:

```
header_size = header.header_size + ( header.header_checksum) << 8)
```

The idea is that for level 0 and level 1, the header size is stored as a `uint8 `and the header_checksum as `uint8`. However, for level2 the header size is stored as a `uint16`, and crc is absent. Instead of parsing the header again using the level 2 definition, we can use this trick. Let's say you have a level 2 header with a header size of `0xffff`, we will have a default lzh header with size `0xff` and checksum `0xff`, so we do `size = 0xff + (0xff << 8) = 0xff`

Reply:
Fixed it and explained the magic trick with an inline comment.

Reply:
Oh, ok! I was just confused about the operator precedence, because of the test files. But I see the testfiles are also changed :smile: